### DISCUSSION TITLE
Unit testing with isar

### DISCUSSION THREAD

**Discussion by flocbit** [isar/isar#230] [Help]
Hi guys,

I am currently struggling with finding a way to properly unit test Isar implementations.

So far I have tried using the test scenarios from `isar-test` package itself (`openTempIsar()`) but got stuck at initializing Isar core in my test environment. My tests fail with the following error message:

`IsarError: Could not initialize IsarCore library. If you create a Flutter app, make sure to add isar_flutter_libs to your dependencies.`

isar_flutter_libs is present in my package.json file, so that should not be the issue.

Does anyone have an idea how to actually use an instance of Isar when testing?

If possible I would like to do it without writing fakes/mocks.

Thanks a lot for any hints!

**Comment by simc** [comment]
To use Isar in Dart (without Flutter), you have to [download the binaries manually](https://github.com/isar/isar/blob/2db8776836bb810752ff1517323ef55d57f6e554/packages/isar_test/tool/setup_tests.sh#L7) and tell Isar where to find them: https://github.com/isar/isar/blob/2db8776836bb810752ff1517323ef55d57f6e554/packages/isar_test/lib/common.dart#L71

**Reply by flocbit** [reply]
@leisim Thank you for the tips, Simon :)

That's what I ended up doing yesterday after I spent quite some time wrapping my head around the current testing environment of Isar. I downloaded Isar repo, cd'd into `isar_test` package, downloaded binaries with some adjustments to the `download_binaries.sh` script (to make it work with my local environment since it's optimised for running in a pipeline I suppose?) and tried to run it locally. I did not manage to get it to run though.

I always received the following error message and could not find a plausible cause for it:

`IsarError: SchemaError: Could not deserialize schema JSON`

I then tried to create a fake database with mockito package. Mocking Isar database object itself was easy, but what I was not able to accomplish was to create a fake IsarCollection without rewriting/stubbing pretty much all of it's properties and functions.

So AFAICS there is currently no way to (easily) test Isar in a Flutter environment neither with running Isar locally nor faking it in an easy way, am I right?

I previously used sembast where they allow to pass an in memory database version for testing purposes. Is something similar feasable for Isar too?

Nevertheless I really appreciate all the work you put in Isar and Hive. It's a delight to work with these packages, so thank you for all your work! 🙇‍♂️

**Reply by Mae623** [reply]
Hello, I came across this message too. I wonder how do you solve this question?
